### PR TITLE
added extra arguments for explicit_auth_flows

### PR DESCRIPTION
**Pull request by hunkeelin**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Related [10958](https://github.com/terraform-providers/terraform-provider-aws/issues/10958)

release notes for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md)

### CODE REVIEW

File: aws/resource_aws_cognito_user_pool_client.go [hunk 1]
```diff
@@ -55,6 +55,12 @@ func resourceAwsCognitoUserPoolClient() *schema.Resource {
 						cognitoidentityprovider.ExplicitAuthFlowsTypeAdminNoSrpAuth,
 						cognitoidentityprovider.ExplicitAuthFlowsTypeCustomAuthFlowOnly,
 						cognitoidentityprovider.ExplicitAuthFlowsTypeUserPasswordAuth,
+						cognitoidentityprovider.ExplicitAuthFlowsTypeAllowAdminUserPasswordAuth,
+						cognitoidentityprovider.AuthFlowTypeAdminUserPasswordAuth,
```
Comment (aaroning) [inline aws/resource_aws_cognito_user_pool_client.go]: You won't need cognitoidentityprovider.AuthFlowTypeAdminUserPasswordAuth, just cognitoidentityprovider.ExplicitAuthFlowsTypeAllowAdminUserPasswordAuth

Comment (hunkeelin) [reply aws/resource_aws_cognito_user_pool_client.go]: done

File: website/docs/r/cognito_user_pool_client.markdown [hunk 2]
```diff
@@ -51,7 +51,7 @@ The following arguments are supported:
 * `allowed_oauth_scopes` - (Optional) List of allowed OAuth scopes (phone, email, openid, profile, and aws.cognito.signin.user.admin).
 * `callback_urls` - (Optional) List of allowed callback URLs for the identity providers.
 * `default_redirect_uri` - (Optional) The default redirect URI. Must be in the list of callback URLs.
-* `explicit_auth_flows` - (Optional) List of authentication flows (ADMIN_NO_SRP_AUTH, CUSTOM_AUTH_FLOW_ONLY, USER_PASSWORD_AUTH).
+* `explicit_auth_flows` - (Optional) List of authentication flows (ADMIN_NO_SRP_AUTH, CUSTOM_AUTH_FLOW_ONLY, USER_PASSWORD_AUTH,ALLOW_ADMIN_USER_PASSWORD_AUTH ,ADMIN_USER_PASSWORD_AUTH,ALLOW_CUSTOM_AUTH,ALLOW_USER_PASSWORD_AUTH,ALLOW_USER_SRP_AUTH,ALLOW_REFRESH_TOKEN_AUTH).
```
Comment (aaroning) [inline website/docs/r/cognito_user_pool_client.markdown]: Remove ADMIN_USER_PASSWORD_AUTH

Comment (hunkeelin) [reply website/docs/r/cognito_user_pool_client.markdown]: > ADMIN_USER_PASSWORD_AUTH
removed

File: website/docs/r/cognito_user_pool_client.markdown [hunk 3]
```diff
@@ -51,7 +51,7 @@ The following arguments are supported:
 * `allowed_oauth_scopes` - (Optional) List of allowed OAuth scopes (phone, email, openid, profile, and aws.cognito.signin.user.admin).
 * `callback_urls` - (Optional) List of allowed callback URLs for the identity providers.
 * `default_redirect_uri` - (Optional) The default redirect URI. Must be in the list of callback URLs.
-* `explicit_auth_flows` - (Optional) List of authentication flows (ADMIN_NO_SRP_AUTH, CUSTOM_AUTH_FLOW_ONLY, USER_PASSWORD_AUTH).
+* `explicit_auth_flows` - (Optional) List of authentication flows (ADMIN_NO_SRP_AUTH, CUSTOM_AUTH_FLOW_ONLY, USER_PASSWORD_AUTH,ALLOW_ADMIN_USER_PASSWORD_AUTH,ALLOW_CUSTOM_AUTH,ALLOW_USER_PASSWORD_AUTH,ALLOW_USER_SRP_AUTH,ALLOW_REFRESH_TOKEN_AUTH).
```
Comment (Haegin) [inline website/docs/r/cognito_user_pool_client.markdown]: Should these have spaces after the commas?

**Comment by bacoboy** [comment]
This appears to have stalled with the holiday.  Any chance this can get some love soon?  Thanks!

**Comment by rednap** [comment]
using a temporary work around with local-exec and aws cli (while we wait):

`resource "null_resource" "add_SRP_to_cognito_app_client" {
    provisioner "local-exec" {
        command = "export AWS_DEFAULT_REGION=\"us-east-1\" && aws cognito-idp update-user-pool-client --user-pool-id ${aws_cognito_user_pool.cog_pool.id} --client-id ${aws_cognito_user_pool_client.cog_client.id} --explicit-auth-flows \"ALLOW_USER_SRP_AUTH\" \"ALLOW_REFRESH_TOKEN_AUTH\" \"ALLOW_CUSTOM_AUTH\""
    }   
    triggers = { 
        build_number = "${timestamp()}"
    }   
    depends_on = [aws_cognito_user_pool.cog_pool, aws_cognito_user_pool_client.cog_client]
}`

every time you apply your user pool will see you changed your auth_flows, update your user pool and client with the old style auth flows again, then your local exec runs and updates them to the new style auth flows. it's not pretty but it works.

File: website/docs/r/cognito_user_pool_client.markdown (hunk 3 above)
Comment (hunkeelin) [reply website/docs/r/cognito_user_pool_client.markdown]: resolved

**Comment by claydanford** [comment]
Any idea when this will be merged? Looks like the commas were resolved. This is blocking work, and I rather not use a custom provider or the cli work around.

Kindly,

CRD

**Comment by claydanford** [comment]
The fix that @rednap created does not work, as it will default all other custom options. For example, updating `explicit_auth_flows` via cli will clear out the `support_identity_providers` configuration.

**Comment by MatusiakMichal** [comment]
Same as @claydanford it indeed is a blocker. May we kindly get this merged already?

**Pull request merged**